server/service/schedule/schedule.go

```diff
@@ -139,126 +153,221 @@ func New(
 //
 // All jobs must be added before calling Start.
 func (s *Schedule) Start() {
-	var m sync.Mutex // protects currentStart and currentWait.
-	currentStart := time.Now()
-	currentWait := 10 * time.Second
+	var intervalStartedAt time.Time // start time of the most recent run of the scheduled jobs
+	var m sync.Mutex                // protects intervalStartedAt
 
-	getWaitTimes := func() (start time.Time, wait time.Duration) {
+	getIntervalStartedAt := func() (start time.Time) {
 		m.Lock()
 		defer m.Unlock()
 
-		return currentStart, currentWait
+		return intervalStartedAt
 	}
 
-	setWaitTimes := func(start time.Time, wait time.Duration) {
+	setIntervalStartedAt := func(start time.Time) {
 		m.Lock()
 		defer m.Unlock()
 
-		currentStart = start
-		currentWait = wait
+		intervalStartedAt = start.Truncate(time.Second)
 	}
 
-	if schedInterval := s.getSchedInterval(); schedInterval < currentWait {
-		setWaitTimes(currentStart, schedInterval)
+	stats, err := s.getStats()
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		level.Debug(s.logger).Log("msg", "start schedule, cron stats not found")
+	case err != nil:
+		level.Error(s.logger).Log("err", "start schedule", "details", err)
+		sentry.CaptureException(err)
+		ctxerr.Handle(s.ctx, err)
+	default:
+		// ok
 	}
 
-	var g sync.WaitGroup
+	if stats == nil {
+		// use empty stats so created and updated timestamps will be the zero time
+		stats = &fleet.CronStats{}
+	}
+	setIntervalStartedAt(stats.CreatedAt)
+
+	initialWait := 10 * time.Second
+	if schedInterval := s.getSchedInterval(); schedInterval < initialWait {
+		initialWait = schedInterval
+	}
+	schedTicker := time.NewTicker(initialWait)
 
-	schedTicker := time.NewTicker(currentWait)
+	var g sync.WaitGroup
 	g.Add(+1)
 	go func() {
-		defer g.Done()
+		defer func() {
+			s.releaseLock()
+			g.Done()
+		}()
 
 		for {
-			_, currWait := getWaitTimes()
-			level.Debug(s.logger).Log("msg", "waiting", "current wait time", currWait)
+			level.Debug(s.logger).Log("waiting", "wait for next tick")
 
 			select {
 			case <-s.ctx.Done():
+				schedTicker.Stop()
 				return
 
 			case <-schedTicker.C:
 				level.Debug(s.logger).Log("waiting", "done")
 
 				schedInterval := s.getSchedInterval()
-				schedTicker.Reset(schedInterval)
-
-				newStart := time.Now()
-				newWait := schedInterval
-				setWaitTimes(newStart, newWait)
 
-				if ok := s.acquireLock(); !ok {
+				stats, err := s.getStats()
+				switch {
+				case errors.Is(err, sql.ErrNoRows):
+					level.Debug(s.logger).Log("msg", "cron stats not found")
+					stats = &fleet.CronStats{}
+					// ok, stats now set to empty so created and updated timestamps will be the zero time
+				case err != nil:
+					level.Error(s.logger).Log("err", "get cron stats", "details", err)
+					sentry.CaptureException(err)
+					ctxerr.Handle(s.ctx, err)
+					// skip ahead to the next interval
+					schedTicker.Reset(schedInterval)
+					continue
+				case stats.Status == fleet.CronStatsStatusPending:
```

Comment:
So the logic here is that if the stats are already pending, it means another instance is still working on it? It feels like we are mixing responsibilities, wouldn't that be the responsibility of the lock? For me cron stats are informational only, their content shouldn't modify what the app does, it's there for humans to reason about in a more human way.

Reply:
Since we're going to know the status information from the stats table anyway, I figured we might as well dump out early and skip the extra call to the lock.

Maybe `cron_stats` isn't quite the right name then. We need a table somewhere to record the status of schedule runs with more granularity than just "locked" or "unlocked" so that the app can coordinate across instances and make decisions about what to do next, how to respond to trigger requests, how to sync intervals, etc.

I think there are three ways we could approach this:
1. Expand the `locks` table to include all the information we need to manage the schedule runs, at which point it looks pretty much like the `cron_stats` table, so we could retain some rows that record prior runs (in which case, it probably makes sense to rename it).
2. Keep the locks table as-is and add a `cron_stats` table (or whatever we want to call it) like in the PR.
3. Expand the `locks` table to manage everything to do with the current run, which would duplicate most of the columns in the `cron_stats` table but with just one row per schedule (again, maybe `locks` isn't the right name any more) AND add another table record all the same information for prior runs.

Reply:
Let's discuss this in the weekly today?